### PR TITLE
sysusers: split up systemd.conf

### LINKED ISSUE #20044: split up systemd/sysusers.d/systemd.conf.in
Hey there.

This is because of https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=990349 .

Right now, Debian creates the various systemd-* users in the maintainer scripts from the respective packages (e.g. systemd, systemd-timesyncd and so on)... (but this might change as Michael Biebl indicated in the bug above).

However, it ships `systemd/sysusers.d/systemd.conf` in the systemd (main) package, so effectively, even if systemd-timesyncd isn't installed, the user still gets created for no reason.
Same with systemd-coredump.

In the above ticket I've asked whether they could simply split up `systemd/sysusers.d/systemd.conf` and ship each part in the respective package, not knowing that this is actually based on this upstream file (`systemd/sysusers.d/systemd.conf.in`).

So Michael asked in the bug, that I might ask systemd upstream, whether they see a point in this and could split up the file, so it would be easier for distros to pick that up.

Thanks,
Chris.

### PR DESCRIPTION
**Pull request by calestyo**
This makes it easier have the respective users/groups only created when their
respective packages are installed.

Fixes #20044.

### CODE REVIEW

File: sysusers.d/systemd-coredump.conf.in [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#  This file is part of systemd.
+#
+#  systemd is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+{% if ENABLE_COREDUMP %}
```
Comment (yuwata) [inline sysusers.d/systemd-coredump.conf.in]: This condition seems not necessary any more.

File: sysusers.d/systemd-network.conf.in [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#  This file is part of systemd.
+#
+#  systemd is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+{% if ENABLE_NETWORKD %}
```
Comment (yuwata) [inline sysusers.d/systemd-network.conf.in]: Also here.

**Review by yuwata — CHANGES_REQUESTED** [review]
I guess conditions in each .conf are not necessary.

File: sysusers.d/systemd-network.conf.in (hunk 2 above)
Comment (mbiebl) [reply sysusers.d/systemd-network.conf.in]: Right, there doesn't need to be a .in pre-processing anymore.

Comment (mbiebl) [reply sysusers.d/systemd-network.conf.in]: argh, for networkd, you still need .in pre-processing for the UID, but no condition, as @yuwata said.

So disregard what I said. I was looking at systemd-oom.conf.in when commenting.

File: sysusers.d/systemd-coredump.conf.in (hunk 1 above)
Comment (calestyo) [reply sysusers.d/systemd-coredump.conf.in]: I've had already asked for it in https://github.com/systemd/systemd/issues/20044#issuecomment-869876506 ...

I guess one can drop it then from `systemd-remote.conf.in`, too? That's why I kept it originally, cause that had it, too.

File: sysusers.d/systemd-network.conf.in (hunk 2 above)
Comment (calestyo) [reply sysusers.d/systemd-network.conf.in]: > argh, for networkd, you still need .in pre-processing for the UID, but no condition, as @yuwata said.

So the `{% if ENABLE_NETWORKD %}` around is needed for the inner `{{SYSTEMD_NETWORK_UID}}` to work?

Comment (keszybz) [reply sysusers.d/systemd-network.conf.in]: No.

**Comment by mbiebl** [comment]
The ubuntu CI failures are expected. If/Once this PR is merged, I'd update the packaging files in the upstream-ci branch to unbreak this again.

**Comment by poettering** [comment]
I am not sure how much I like this. Splitting this up means people's overrides and extends don't work anymore. The file names this stuff is listed in kinda is an API

**Comment by poettering** [comment]
What's the benefit of this? people being allergic to a bunch of extra UIDs being allocate? is that really something the trouble of breaking API for? Admittedly it's a weak for API, but it's effectively API nonetheless, and we *did* have reports about this when we did something similar in earlier cases.

**Comment by mbiebl** [comment]
> What's the benefit of this?

I'm a bit torn on this. My first reaction was similar to yours. Contemplating a bit more on it, I see two possible benefits:
- For distros that split up systemd, it's as @keszybz said, more elegant / conceptually cleaner that the invididual package declares what user it needs/uses. 
- It makes it easier to override/extend individual entries. If you have a single/monolithic systemd.conf and you only want to change the settings for say systemd-network, you'll have to override all of systemd.conf. This means, whenever systemd.conf changes, you have to review those changes and adapt accordingly if necessary.

You raise a valid concern though, that it might break existing systems, which have an override for the monolithic systemd.conf.

At least for Debian/Ubuntu, we could cover that, by adding an appropriate preinst check. If such an `/etc/sysusers.d/systemd.conf` exists, we could output a warning.

**Comment by mbiebl** [comment]
> Admittedly it's a weak for API, but it's effectively API nonetheless, and we _did_ have reports about this when we did something similar in earlier cases.

Maybe I missed that, but what was this about exactly? Maybe this is something we can address in our Debian packages with a corresponding upgrade check.

**Comment by keszybz** [comment]
The "api break" is negligible. The easiest and most obvious way to "override" the sysusers settings is to precreate the user and/or group. I think it's very unlikely that this would create any problems. We did such splits with tmpfiles before, and there it's more likely for people to create overrides, and I don't recall any complains. We'd just add a NEWS entry and things will be fine.

**Comment by calestyo** [comment]
I've updated the first commit to remove the meson stuff, and added a 2nd one which also removes it for the already separate `systemd-remote.conf.in`.

Felt that didn't belong into one commit, as for `systemd-remote.conf.in` it's not really about splitting the file up, but just removing what was (AFAIU) already not needed without the split.

**Comment by yuwata** [comment]
So, now, some of them are not necessary to be processed with jinja2, and `.in` suffix can be dropped.

**Comment by calestyo** [comment]
> So, now, some of them are not necessary to be processed with jinja2, and `.in` suffix can be dropped.

updated accordingly

**Comment by yuwata** [comment]
Thank you for the quick update. But meson file also needs to be updated to handle files without the suffix correctly.

**Comment by calestyo** [comment]
I've never really used meson... how do I conditionally include the non-`.in` files? Or at least I think they should still be only included if the respective flag is set?

**Comment by keszybz** [comment]
>  how do I conditionally include the non-`.in` files? Or at least I think they should still be only included if the respective flag is set?

The `tmpfiles.d/meson.build` file has a good example.

**Comment by calestyo** [comment]
Sorry, for needing help again, but when I closely follow what's done for tmpfiles.d, e.g. something like:
```
files = [['README'], ''],
         ['systemd-coredump.conf', 'ENABLE_COREDUMP'],
         ['systemd-oom.conf',      'ENABLE_OOMD'],
         ['systemd-remote.conf',   'ENABLE_REMOTE']]

foreach pair : files
        if not enable_sysusers
                # do nothing
        elif pair[1] == '' or conf.get(pair[1]) == 1
                install_data(pair[0], install_dir : sysusersdir)
        else
                message('Not installing sysusers.d/@0@ because @1@ is @2@'
                        .format(pair[0], pair[1], conf.get(pair[1], 0)))
        endif
endforeach
```

Than I obviously have the problem that I cannot really fit the `and conf.get('HAVE_MICROHTTPD') == 1]]` into it.
Is there either some nice way in meson to make the 2nd element of the pair another array that's iterated over, or shall I simply included the boolean outcome of the condition in `files` and take that?

**Comment by keszybz** [comment]
> shall I simply included the boolean outcome of the condition in `files` and take that?

Since there's just one file, you can do something like this:
```diff
  if enable_sysusers
-         install_data('README', install_dir : sysusersdir)
+         install_data('README',
+                      'systemd-remote.conf',
+                      install_dir : sysusersdir)
  endif
```

**Comment by calestyo** [comment]
Okay check please check these commits if they fit.

**Comment by calestyo** [comment]
Anything left to do on this?

**Comment by keszybz** [comment]
Sorry for the delay. Looks great, thanks.

**Pull request merged**